From 534de6fa22d036541181528126dd93420801b499 Mon Sep 17 00:00:00 2001
From: Satish Kumar Nagireddy <satish.nagireddy.nagireddy@xilinx.com>
Date: Wed, 27 Mar 2019 18:03:31 -0700
Subject: [PATCH 0584/1851] v4l: xilinx: scd: Use correct threshold value

commit b8ecd1a2f188e84577302c5e2fd4590391174ec1 from
https://github.com/Xilinx/linux-xlnx.git

The driver is using threshold value as 1 and it is 0.5 according to the
specification. This patch uses the right threshold value.

Reference (page 21):
https://www.xilinx.com/support/documentation/ip_documentation/v_scenechange/v1_0/pg322-v-scenechange-detect.pdf

Signed-off-by: Satish Kumar Nagireddy <satish.nagireddy.nagireddy@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 .../media/platform/xilinx/xilinx-scenechange-channel.c | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/drivers/media/platform/xilinx/xilinx-scenechange-channel.c b/drivers/media/platform/xilinx/xilinx-scenechange-channel.c
index 37b2eb144c2e..6d95e870214b 100644
--- a/drivers/media/platform/xilinx/xilinx-scenechange-channel.c
+++ b/drivers/media/platform/xilinx/xilinx-scenechange-channel.c
@@ -55,6 +55,8 @@
 
 #define XSCD_V_SUBSAMPLING		16
 #define XSCD_BYTE_ALIGN			16
+#define MULTIPLICATION_FACTOR		100
+#define SCENE_CHANGE_THRESHOLD		0.5
 
 #define XSCD_SCENE_CHANGE		1
 #define XSCD_NO_SCENE_CHANGE		0
@@ -342,14 +344,16 @@ static const struct media_entity_operations xscd_media_ops = {
 static void xscd_event_notify(struct xscd_chan *chan)
 {
 	u32 *eventdata;
-	u32 sad;
+	u32 sad, scd_threshold;
 
 	sad = xscd_read(chan->iomem, XSCD_SAD_OFFSET +
 			(chan->id * XILINX_XSCD_CHAN_OFFSET));
-	sad = (sad * 16) / (chan->format.width * chan->format.height);
+	sad = (sad * XSCD_V_SUBSAMPLING * MULTIPLICATION_FACTOR) /
+	       (chan->format.width * chan->format.height);
 	eventdata = (u32 *)&chan->event.u.data;
+	scd_threshold = SCENE_CHANGE_THRESHOLD * MULTIPLICATION_FACTOR;
 
-	if (sad >= 1)
+	if (sad > scd_threshold)
 		eventdata[0] = XSCD_SCENE_CHANGE;
 	else
 		eventdata[0] = XSCD_NO_SCENE_CHANGE;
-- 
2.31.1

